Src/VimWpf/Implementation/ImeCoordinator/ImeCoordinator.cs

```diff
@@ -0,0 +1,423 @@
+{
+    [Export(typeof(IVimBufferCreationListener))]
+    internal sealed class ImeCoordinator : IVimBufferCreationListener, IDisposable
+    {
+        internal enum InputMode
+        {
+            None = 0,
+            Command = 1,
+            Insert = 2,
+            Search = 3,
+        };
+
+        /// <summary>
+        /// A mapping from input mode to IME state using the
+        /// global IME-related settings as a backing store
+        /// </summary>
+        internal class InputModeState
+        {
+            private readonly IVimGlobalSettings _globalSettings;
+
+            private bool _synchronizingSettings;
+
+            public InputModeState(IVimGlobalSettings globalSettings)
+            {
+                _globalSettings = globalSettings;
+
+                _synchronizingSettings = false;
+            }
+
+            public bool SynchronizingSettings
+            {
+                get
+                {
+                    return _synchronizingSettings;
+                }
+            }
+
+            public InputMethodState this[InputMode inputMode]
+            {
+                get
+                {
+                    return GetState(inputMode);
+                }
+                set
+                {
+                    SynchronizeState(inputMode, value);
+                }
+            }
+
+            private InputMethodState GetState(InputMode inputMode)
+            {
+                switch (inputMode)
+                {
+                    case InputMode.Command:
+                        if (_globalSettings.ImeCommand)
+                        {
+                            return GetState(InputMode.Insert);
+                        }
+                        else
+                        {
+                            return InputMethodState.Off;
+                        }
+
+                    case InputMode.Insert:
+                        if (_globalSettings.ImeInsert == 2)
+                        {
+                            return InputMethodState.On;
+                        }
+                        else
+                        {
+                            return InputMethodState.Off;
+                        }
+
+                    case InputMode.Search:
+                        if (_globalSettings.ImeSearch == -1)
+                        {
+                            return GetState(InputMode.Insert);
+                        }
+                        else
+                        {
+                            if (_globalSettings.ImeSearch == 2)
+                            {
+                                return InputMethodState.On;
+                            }
+                            else
+                            {
+                                return InputMethodState.Off;
+                            }
+                        }
+
+                    default:
+                        throw new ArgumentException("inputMode");

```

Comment:
Can use `nameof(inputMode)`

Reply:
I'm still catching up to now ancient C# features!